src/classification/mondrian_tree.rs

```diff
@@ -0,0 +1,373 @@
+                    "Non-root node has no parent"
+                )
+            }
+        }
+
+        let children_l: Vec<usize> = self.nodes.iter().filter_map(|node| node.left).collect();
+        let children_r: Vec<usize> = self.nodes.iter().filter_map(|node| node.right).collect();
+        let children = [children_l.clone(), children_r.clone()].concat();
+        let mut seen = HashSet::new();
+        let has_duplicates = children.iter().any(|item| !seen.insert(item));
+        assert!(
+            !has_duplicates,
+            "Multiple nodes share one child. Children left: {:?}, Children right: {:?}",
+            children_l, children_r
+        );
+    }
+
+    fn compute_split_time(
+        &self,
+        time: F,
+        exp_sample: F,
+        node_idx: usize,
+        y: usize,
+        extensions_sum: F,
+    ) -> F {
+        if self.nodes[node_idx].is_dirac(y) {
+            // println!(
+            //     "go_downwards() - node: {node_idx} - extensions_sum: {:?} - all same class",
+            //     extensions_sum
+            // );
+            return F::zero();
+        }
+
+        if extensions_sum > F::zero() {
+            let split_time = time + exp_sample;
+
+            // From River: If the node is a leaf we must split it
+            if self.nodes[node_idx].is_leaf {
+                // println!(
+                //     "go_downwards() - node: {node_idx} - extensions_sum: {:?} - split is_leaf",
+                //     extensions_sum
+                // );
+                return split_time;
+            }
+
+            // From River: Otherwise we apply Mondrian process dark magic :)
+            // 1. We get the creation time of the childs (left and right is the same)
+            let child_idx = self.nodes[node_idx].left.unwrap();
+            let child_time = self.nodes[child_idx].time;
+            // 2. We check if splitting time occurs before child creation time
+            if split_time < child_time {
+                // println!(
+                //     "go_downwards() - node: {node_idx} - extensions_sum: {:?} - split mid tree",
+                //     extensions_sum
+                // );
+                return split_time;
+            }
+            // println!("go_downwards() - node: {node_idx} - extensions_sum: {:?} - not increased enough to split (mid node)", extensions_sum);
+        } else {
+            // println!(
+            //     "go_downwards() - node: {node_idx} - extensions_sum: {:?} - not outside box",
+            //     extensions_sum
+            // );
+        }
+
+        F::zero()
+    }
+
+    fn go_downwards(&mut self, node_idx: usize, x: &Array1<F>, y: usize) -> usize {
+        let time = self.nodes[node_idx].time;
+        let node_min_list = &self.nodes[node_idx].min_list;
+        let node_max_list = &self.nodes[node_idx].max_list;
+        let extensions = {
+            let e_min = (node_min_list - x).mapv(|v| F::max(v, F::zero()));
```

Comment:
I am still trying to grasp a basic understanding of Rust, so pardon me if I am getting this wrong. Should you use F::min instead?